Sming/system/flashmem.c

```diff
@@ -1,102 +1,205 @@
-  const uint32_t blkmask = INTERNAL_FLASH_WRITE_UNIT_SIZE - 1;
-
-  // Align the start
-  if( toaddr & blkmask )
-  {
-    rest = toaddr & blkmask;
-    temp = toaddr & ~blkmask; // this is the actual aligned address
-    flashmem_read_internal( tmpdata, temp, blksize );
-    for( i = rest; size && ( i < blksize ); i ++, size --, pfrom ++ )
-      tmpdata[ i ] = *pfrom;
-    flashmem_write_internal( tmpdata, temp, blksize );
-    if( size == 0 )
-      return ssize;
-    toaddr = temp + blksize;
-  }
-  // The start address is now a multiple of blksize
-  // Compute how many bytes we can write as multiples of blksize
-  rest = size & blkmask;
-  temp = size & ~blkmask;
-  // Program the blocks now
-  if( temp )
-  {
-	flashmem_write_internal( pfrom, toaddr, temp );
-    toaddr += temp;
-    pfrom += temp;
-  }
-  // And the final part of a block if needed
-  if( rest )
-  {
-    flashmem_read_internal( tmpdata, toaddr, blksize );
-    for( i = 0; size && ( i < rest ); i ++, size --, pfrom ++ )
-      tmpdata[ i ] = *pfrom;
-    flashmem_write_internal( tmpdata, toaddr, blksize );
-  }
-  return ssize;
+	return (a < b) ? a : b;
 }
 
-uint32_t flashmem_read( void *to, uint32_t fromaddr, uint32_t size )
+// For release builds, bypass internal flash read routines and go direct to SDK
+#ifdef DEBUG_BUILD
+
+#define FLASHMEM_READ flashmem_read_internal
+#define FLASHMEM_WRITE flashmem_write_internal
+
+#else
+
+static inline uint32_t _flashmem_write_internal(const void* from, uint32_t toaddr, uint32_t size)
 {
-  uint32_t temp, rest, ssize = size;
-  unsigned i;
-  char tmpdata[ INTERNAL_FLASH_READ_UNIT_SIZE ];
-  uint8_t *pto = ( uint8_t* )to;
-  const uint32_t blksize = INTERNAL_FLASH_READ_UNIT_SIZE;
-  const uint32_t blkmask = INTERNAL_FLASH_READ_UNIT_SIZE - 1;
-
-  // Align the start
-  if( fromaddr & blkmask )
-  {
-    rest = fromaddr & blkmask;
-    temp = fromaddr & ~blkmask; // this is the actual aligned address
-    flashmem_read_internal( tmpdata, temp, blksize );
-    for( i = rest; size && ( i < blksize ); i ++, size --, pto ++ )
-      *pto = tmpdata[ i ];
-
-    if( size == 0 )
-      return ssize;
-    fromaddr = temp + blksize;
-  }
-  // The start address is now a multiple of blksize
-  // Compute how many bytes we can read as multiples of blksize
-  rest = size & blkmask;
-  temp = size & ~blkmask;
-  // Program the blocks now
-  if( temp )
-  {
-	flashmem_read_internal( pto, fromaddr, temp );
-    fromaddr += temp;
-    pto += temp;
-  }
-  // And the final part of a block if needed
-  if( rest )
-  {
-	flashmem_read_internal( tmpdata, fromaddr, blksize );
-    for( i = 0; size && ( i < rest ); i ++, size --, pto ++ )
-      *pto = tmpdata[ i ];
-  }
-  return ssize;
+	SpiFlashOpResult r = spi_flash_write(toaddr, (uint32*)from, size);
+	return (SPI_FLASH_RESULT_OK == r) ? size : 0;
 }
 
-bool flashmem_erase_sector( uint32_t sector_id )
+static inline uint32_t _flashmem_read_internal(void* to, uint32_t fromaddr, uint32_t size)
```

Comment:
`_flashmem_read_internal` -> `flashmem_read_internal`. Make appropriate fixes If there are also other places in this PR where the name of the function ot the variables start or end with underscore.

Reply:
OK, but we'll have to leave `_flash_code_end` though

Reply:
> OK, but we'll have to leave _flash_code_end though

Yes, that's correct.